### PR TITLE
esteban kocian feat(IntegrationTests): add first scenarios for integrations tests. mecommerce/feature/CXL-23

### PR DESCRIPTION
**Pull request by estebankocian**
# Overview of the changes
Add modules needed for integration tests creation. Includes currently supported and un-supported scenarios:

Supported
- fullOneItemRefundWithTaxesWithShipping,
- fullTwoItemsSameSkuRefundWithTaxesWithShipping,
- fullTwoItemsDifferentSkuRefundWithTaxesWithShipping,
- partialTwoItemsSameSkuRefundWithTaxesWithShipping,
- partialTwoItemsDifferentSkuRefundWithTaxesWithShipping,
- fullMoreThanTwoItemsDifferentSkuRefundWithTaxesWithShipping,
- fullMoreThanTwoItemsSameSkuRefundWithTaxesWithShipping,

Not supported
- fullOneItemRefundWithDiscount,
- fullTwoItemsRefundWithDiscount,
- partialMoreThanTwoItemsRefund,
- fullMoreThanTwoItemsRefundWithDiscount,

### Jira Link
https://thirdlove.atlassian.net/browse/CXL-23

# Steps for testing
1. checkout the branch on your local env
2. Select the tests that you want to include to be run by setting the skip property present on the description cases files (i.e. test/integration/scenarios/test-cases-*)
2. run the script by `npm run test:integration`.
3. Also checkout a new branch with rc/* and check that the CI run the test and work properly.

# FYI / Please Review:
There is a complementary [document](https://thirdlove.atlassian.net/wiki/spaces/TLENG/pages/779289604) on Confluence about the covered Scenarios

### CODE REVIEW

File: src/clients/shopify.js [hunk 1]
```diff
@@ -138,10 +138,133 @@ const calculateRefund = async (orderId, calculateRequest) => (
...
 );
 
+/**
+ * Create a new Shopify Order
+ * @param {object} orderCreateRequest
+ * @return {object}
+ */
+const createShopifyOrder = async orderCreateRequest => (
+  post(
+    `/admin/api/${CONFIG_SHOPIFY_API_VERSION}/orders.json`,
```
Comment (nvalentinis-git) [inline src/clients/shopify.js]: I have refactorized this, now the this URL is defined on the Fetch base URL

File: src/modules/shopify-logic.js [hunk 2]
```diff
@@ -224,6 +224,7 @@ function refundOrder(shopifyOrder, lineItems) {
   };
   return shopifyClient.calculateRefund(orderId, calculateRefundBody)
     .then((response) => {
+      logger.debug('CalculateRefund: ', { response });
```
Comment (nvalentinis-git) [inline src/modules/shopify-logic.js]: Do we need this logger? Keep in mind we can log sensitive information

File: src/process/refunds/main-process.js [hunk 3]
```diff
@@ -80,6 +80,23 @@ function generateCustomReturnReport() {
   generateReport(email);
 }
 
+/**
+ * Method used to run the process with parametrized values when used on integrations tests.
+ *
+ * [FILE_NAME]: XLSX File path to be processed
+ * [Fontana|Greenwood]: String indicating the Fedex store
+ *
+ * ie: $npm run customFedexReturnReport [DATE] [FILE_NAME] [Fontana|Greenwood]
```
Comment (nvalentinis-git) [inline src/process/refunds/main-process.js]: This name has changed on this project

File: test/integration/scenarios/refunds.int.js [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  data.unshift(getXlsFedexSheet(), getXlsFedexSheet());
+  const wsName = 'integrationTests';
+  const wb = XLSX.utils.book_new();
+  const ws = XLSX.utils.json_to_sheet(data, { cellDates: true });
+  XLSX.utils.book_append_sheet(wb, ws, wsName);
+  await XLSX.writeFile(wb, fileName);
+};
+
+const refundsTest = () => describe('Create orders and run auto-return process in order to check if automated return is possible', () => {
+  cases('Should create an order and the associated xls for the returned items and run customFedexReturnReport script',
```
Comment (nvalentinis-git) [inline test/integration/scenarios/refunds.int.js]: The name of the process `customFedexReturnReport ` has changed

File: test/integration/scenarios/refunds.int.js [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+      expect(orderRefunded.order).toBeDefined();
+      const { refunds } = orderRefunded.order;
+      logger.info('Order Refunded with detail:', { orderRefunded });
+      if (opt.expectedToBeRefunded) {
+        expect(refunds.length).toBeGreaterThan(0);
+      } else {
+        expect(refunds.length).toEqual(0);
+      }
+    }, {
+      // supported
```
Comment (nvalentinis-git) [inline test/integration/scenarios/refunds.int.js]: Minor: I would put `//Scenarios that should be automated`

File: test/integration/scenarios/refunds.int.js [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+      fullOneItemRefundWithTaxesWithShipping,
+      fullTwoItemsSameSkuRefundWithTaxesWithShipping,
+      fullTwoItemsDifferentSkuRefundWithTaxesWithShipping,
+      partialTwoItemsSameSkuRefundWithTaxesWithShipping,
+      partialTwoItemsDifferentSkuRefundWithTaxesWithShipping,
+      fullMoreThanTwoItemsDifferentSkuRefundWithTaxesWithShipping,
+      fullMoreThanTwoItemsSameSkuRefundWithTaxesWithShipping,
+      // not supported
```
Comment (nvalentinis-git) [inline test/integration/scenarios/refunds.int.js]: Minor: I would put `//Scenarios that should not be automated`

File: test/integration/secuential-integration-test.spec.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  refundsTest,
+} = require('./scenarios');
+
+const {
+  getJestTimeout,
+} = require('./utils');
+
+jest.setTimeout(getJestTimeout());
+
+describe('Parse Integration Tests sequentially run', () => {
```
Comment (nvalentinis-git) [inline test/integration/secuential-integration-test.spec.js]: I would change the name to something more representative of the Test `Refund Process Integration Test sequentially `

File: test/integration/scenarios/test-cases-refunds-not-supported.js [hunk 8]
```diff
@@ -0,0 +1,193 @@
+// must be not automated with reason: "Order with discounts"
+const fullOneItemRefundWithDiscount = {
+  skip: true,
```
Comment (esoliani) [inline test/integration/scenarios/test-cases-refunds-not-supported.js]: does this variable used?

File: test/integration/scenarios/test-cases-refunds-not-supported.js [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+  expectedToBeRefunded: false,
+  // data related to order
+  province: 'California',
```
Comment (esoliani) [inline test/integration/scenarios/test-cases-refunds-not-supported.js]: we could group some attributes as address, tags, lineItems. here and in the rest of the objects
```
{
  ...californiaAddress,
  ...rirrTag,
  ...lineItems,
}
```

File: src/clients/shopify.js (hunk 1 above)
Comment (estebankocian) [reply src/clients/shopify.js]: 6b7db7c1806947e45372cdd244541b232d7c38e8

File: src/modules/shopify-logic.js (hunk 2 above)
Comment (estebankocian) [reply src/modules/shopify-logic.js]: 6b7db7c1806947e45372cdd244541b232d7c38e8

File: src/process/refunds/main-process.js (hunk 3 above)
Comment (estebankocian) [reply src/process/refunds/main-process.js]: 6b7db7c1806947e45372cdd244541b232d7c38e8

File: test/integration/scenarios/refunds.int.js (hunk 4 above)
Comment (estebankocian) [reply test/integration/scenarios/refunds.int.js]: 6b7db7c1806947e45372cdd244541b232d7c38e8

File: test/integration/scenarios/refunds.int.js (hunk 5 above)
Comment (estebankocian) [reply test/integration/scenarios/refunds.int.js]: 6b7db7c1806947e45372cdd244541b232d7c38e8

File: test/integration/scenarios/refunds.int.js (hunk 6 above)
Comment (estebankocian) [reply test/integration/scenarios/refunds.int.js]: 6b7db7c1806947e45372cdd244541b232d7c38e8

File: test/integration/secuential-integration-test.spec.js (hunk 7 above)
Comment (estebankocian) [reply test/integration/secuential-integration-test.spec.js]: 6b7db7c1806947e45372cdd244541b232d7c38e8

File: test/integration/scenarios/refunds.int.js [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+  cases('Should create an order and the associated xls for the returned items and run customFedexReturnReport script',
+    async (opt) => {
+      // create the order with a sale transaction which means
+      // authorized and captured ready to be refunded
+      const orderCreated = await createOrder(opt);
+      expect(orderCreated.order.name).toBeDefined();
+      logger.info(`Shopify Order created with name: ${orderCreated.order.name}`);
+
+      // fulfill the order
+      const orderFulfillment = await createFulfillment(orderCreated, opt);
```
[the remark this answers is not in the document]
Comment (estebankocian) [reply test/integration/scenarios/refunds.int.js]: 6b7db7c1806947e45372cdd244541b232d7c38e8

File: test/integration/scenarios/test-cases-refunds-not-supported.js (hunk 8 above)
Comment (estebankocian) [reply test/integration/scenarios/test-cases-refunds-not-supported.js]: yes. is used by jest to exclude cases.

File: test/integration/scenarios/test-cases-refunds-not-supported.js (hunk 9 above)
Comment (estebankocian) [reply test/integration/scenarios/test-cases-refunds-not-supported.js]: 6b7db7c1806947e45372cdd244541b232d7c38e8

File: src/clients/shopify.js [hunk 11]
```diff
@@ -60,7 +63,7 @@ const call = async (method, url, requestPayload) => {
  * @return The data {Object}
  */
 const getProducts = async productIds => (
-  call('GET', `/products.json?ids=${productIds.join(',')}`)
+  call(METHOD_GET, `/products.json?ids=${productIds.join(',')}`)
```
Comment (maborri) [inline src/clients/shopify.js]:
```suggestion
  call(METHOD_GET, `/products.json?ids=${productIds.join(',')}`)
```
call(METHOD_GET, `/products.json?ids=${productIds.join()}`)

Comment (maborri) [inline src/clients/shopify.js]:
```suggestion
  call(METHOD_GET, `/products.json?ids=${productIds.join()}`)
```

**Pull request merged**